server/sonar-server/src/main/java/org/sonar/server/computation/step/FillMeasuresWithVariationsStep.java

```diff
@@ -0,0 +1,234 @@
+import javax.annotation.Nullable;
+import org.sonar.core.measure.db.PastMeasureDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.measure.Measure;
+import org.sonar.server.computation.measure.MeasureKey;
+import org.sonar.server.computation.measure.MeasureRepository;
+import org.sonar.server.computation.measure.MeasureVariations;
+import org.sonar.server.computation.metric.Metric;
+import org.sonar.server.computation.metric.MetricRepository;
+import org.sonar.server.computation.period.Period;
+import org.sonar.server.computation.period.PeriodsHolder;
+import org.sonar.server.db.DbClient;
+
+import static org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor.Order.PRE_ORDER;
+
+/**
+ * Set variations on all numeric measures found in the repository.
+ * This step MUST be executed after all steps that create some measures
+ *
+ * Note that measures on developer are not handle yet.
+ */
+public class FillMeasuresWithVariationsStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodsHolder periodsHolder;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  private final Function<PastMeasureDto, MeasureKey> pastMeasureToMeasureKey;
+
+  public FillMeasuresWithVariationsStep(DbClient dbClient, TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder, final MetricRepository metricRepository,
+    MeasureRepository measureRepository) {
+    this.dbClient = dbClient;
+    this.treeRootHolder = treeRootHolder;
+    this.periodsHolder = periodsHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+
+    this.pastMeasureToMeasureKey = new Function<PastMeasureDto, MeasureKey>() {
+      @Nullable
+      @Override
+      public MeasureKey apply(@Nonnull PastMeasureDto input) {
+        Metric metric = metricRepository.getById(input.getMetricId());
+        return new MeasureKey(metric.getKey(), input.getCharacteristicId(), input.getRuleId());
+      }
+    };
+  }
+
+  @Override
+  public void execute() {
+    final DbSession dbSession = dbClient.openSession(false);
+    try {
+      final Iterable<Metric> metrics = FluentIterable.from(metricRepository.getAll()).filter(NumericMetric.INSTANCE);
+      final List<Integer> metricIds = FluentIterable.from(metrics).transform(MetricDtoToMetricId.INSTANCE).toList();
+      new VariationMeasuresVisitor(dbSession, metricIds).visit(treeRootHolder.getRoot());
+    } finally {
+      dbSession.close();
+    }
+  }
+
+  private class VariationMeasuresVisitor extends DepthTraversalTypeAwareVisitor {
+
+    private final DbSession session;
+    private final List<Integer> metricIds;
```

Comment:
why don't use use a Map<String, Metric> of Metric by key next to this List of metrid ids. This would save having to call `MeasureRepository.getByKey` in `setVariationMeasure`